### ISSUE TITLE
[Design] - Rate Limit, Alerts, Monitoring, QA, Deployment Steps

### ISSUE THREAD

**Issue by andreabadesso** [HathorNetwork/hathor-wallet-service-old#80]
# CI

- [x] We need to run tests on every push
    - [x] To be able to do this, we need to spawn a MySQL on CI
    - [x] To be able to do this, we need to spawn a Redis server on CI
    - [x] To be able to do this, we need migrations on at least a script to initialize the database
    
PR: https://github.com/HathorNetwork/hathor-wallet-service/pull/97

# CD and IaaS

This is being worked on by @luislhl on https://github.com/HathorNetwork/ops-tools/issues/34

# Database

* We currently don't have migrations, that's necessary for automated deploys.
* We can deploy the database from the last snapshot so we don't have to sync from genesis on every deploy

## Snapshots

Snapshot daily

PR: https://github.com/HathorNetwork/ops-tools/pull/58

# Monitoring

There are some critical events that we need to be alerted of:

**Wallet Service**
- [ ] Error on balance calculation
- [ ] Error on MySQL connection
- [ ] Error on FullNode connection
- [ ] Wether the wallet-service is being properly synced with the full-node in realtime 
- [ ] Monitor database metrics

**Daemon**
- [ ] A reorg is detected with more than 1000 blocks difference
- [ ] More than X minutes/seconds without a new block from the connected fullnode
- [ ] Websocket connection lost with the full-node after X retries
- [ ] Error on Wallet Service connection

**Comment by luislhl** [comment]
# Continuous Deployment

We already have migrations implemented, so a automated CD pipeline is already possible.

However, we need to discuss how to make everything work together, given that:

1. We were planning to not let the Database publicly accessible
2. By trying to access the database from Github Actions, we have no control over the source IP that this access comes from.

These two facts make it harder for us to trigger the migrations directly from Github Actions.
I have come up with some options to solve this problem:

### Give Github Actions AWS permissions to update the Security Group
The idea is that we would get the IP of the current runner, then update the Security Group to allow access from it, run the migrations, and finally update the Security Group again removing the allowed IP.

The only security concern I see here is that of making sure no one we don't trust have write access to the repository where those secrets are set.

Since the IP would be allowed only during a short time span, I don't think this would open any significant breaches.

### Use Self-hosted runners
https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners#about-self-hosted-runners

This is possibly the most secure option, since we would run our own runner, that would be used by our Action.
However, there is more work involved in making this work (I don't know how much yet), and more maintenance work to keep the machine updated.

This solves our problem because we could just allow our own runner's IP to access the database, without having to provide AWS secrets to the Action.

### Create a new Lambda that runs migrations
In this option, Github Actions would still need to have AWS credentials that allowed it to call this Lambda.

Then, we would allow the Lambda to access the database.

One problem that may arise is if some migration in the future takes more than 15 minutes to run, which would hit the Lambda timeout. Depending on what the migration does, this is not so hard to happen in large tables. In Github Actions jobs, this limit is 6 hours.

### Use AWS CodePipeline instead of Github Actions
It's capable of running jobs inside the VPC: https://aws.amazon.com/blogs/devops/access-resources-in-a-vpc-from-aws-codebuild-builds/

It costs $1/month for each active pipeline, so not too much, and is capable of integrating with Github.

It seems like a very good option for us, we just need a little extra time to configure it.

## My Suggestion
I think we should give AWS CodePipeline a try. It seems the best option for use-cases like this where we have to deploy things accessible only inside the VPC.

If it works well for WalletService, we could define it as our new standard for deployments, and leave Github Actions only with validation tasks, or other tasks that are better to stay publicly accessible.

**Comment by luislhl** [comment]
# Monitoring

Suggestions on how we could monitor the different events we are intending to monitor.

### Wallet Service

> Monitor database metrics

It would be easy to create alarms in CloudWatch for database metrics.

This is already being done in:
https://github.com/HathorNetwork/ops-tools/pull/59

>  Error on balance calculation
>  Error on MySQL connection
>  Error on FullNode connection

 For those, we just need to make sure the Lambdas are exiting with an error status in all the cases, then we can create an Alert in AWS triggered by errors in the Lambdas.
 
 The only problem I see with this is how we will make sure new Lambdas are also included in this monitoring.
 
>  Wether the wallet-service is being properly synced with the full-node in realtime

We have some options here that we should evaluate.

The first one is to simply create an alert in AWS to warn us in case the Lambda `onNewTxRequest` stops being called regularly, or in case it is erroring out. This would be easy to do and to send the alerts to Slack using AWS Chatbot + CloudWatch alerts.

Another option suggested by @andreabadesso is to export the highest height received by the WalletService to Prometheus, and then have an alert to compare it against the height in our FullNodes.

This is a good suggestion, since it could also catch cases where the sync is happening, but is just not fast enough. However, it brings additional complexity that we would need to solve.

We would need a way to expose the metrics to Prometheus, which in a Lambda environment is a little more challenging. The alternatives would be:
- Have a `/metrics` endpoint in API Gateway, which queries directly in MySQL for the height. This query is currently taking circa 1.5s to run. If we need to add other metrics, they would add other queries.
- Somehow share state between different Lambdas. Maybe in S3 or other way. Then we would make one Lambda update the file on specific events, like new blocks or transactions, and the other would just return this in the `/metrics` endpoint. But this doesn't feel like a good solution.
- Install a Prometheus PushGateway, and make the Lambdas push metrics there. This seems the most future-proof solution. However, since we are in the process of migration to a new Grafana/Prometheus installation, ideally we would install this PushGateway in this new installation. But our main full-nodes are on the old one, and the WalletService full-node is not monitored yet.
   
 
### Daemon

> A reorg is detected with more than 1000 blocks difference

This is an event that, as far as I understand, we are only capable of detecting inside the Daemon logic. So the Daemon needs a way to warn some other tool when this happens.

Again, some options. 

1. Just use logs with specific markers (like including an `[ALERT]` string), and send them to CloudWatch. Then, we would be able to create alerts based on the log content.
    - The first problem here is that we are not sending the Kubernetes logs yet to CloudWatch.
    - The second problem is that we would be unable to send the log message together with the alert. To do this, additional effort would be needed: https://stackoverflow.com/questions/60171078/how-to-send-alert-based-on-log-message-on-cloudwatch/60171655?noredirect=1#comment120741778_60171655
    
2. Make the daemon send messages directly to AWS SNS, then route then to AWS Chatbot/Slack. Or send to a service of ours, that would handle them. But in both cases we would need to make the daemon aware of the solution.

3. Make the daemon log the alerts to some file, then have another container read those logs, and for each entry, send an alert somewhere. This container could be built as a reusable container. Everytime we had a similar case, we would use it.

> More than X minutes/seconds without a new block from the connected fullnode

 Isn't this solved by the alert in WalletService synx with full-nodes?

> Websocket connection lost with the full-node after X retries

This is similar to the reorg case. The daemon needs a way to push an alert somewhere warning about this.

 > Error on Wallet Service connection:

 Isn't this solved by alerting on Lambda errors and timeouts? 
 
## My Suggestion
### Wallet Service
In a first moment, use just AWS to monitor errors in Lambdas and invocations of `onNewTxRequest` to make sure the sync is working. This already provides us with a good level of reliability.

Then, we would have time to:
1. Finish the Prometheus/Grafana migraiton
2. Install PushGateway and make a PoC of Lambdas pushing metrics to it. We would create a new issue for this to be prioritized in the future.

### Daemon
I like the option of sending the logs to CloudWatch, and this is something we would have to solve anyway at some point.

So, I think we should solve it and then create the alerts there in case the logs contain some string. Then, the Daemon just have to include the string in every log it wants to trigger an alert for.

I think it will be enough to not have the log message in the alert for now. We can go check the log message in CloudWatch everytime an alert arrives, to understand what exactly happened.

But an issue can be created to track this improvement of also alerting the log message.

**Comment by luislhl** [comment]
# Rate Limits

@andreabadesso and @pedroferreira1 have agreed that we could start with simple rate-limits on endpoints, and create another issue to discuss more complex strategies.

So, the idea for now is to just have them configured in API Gateway through Serverless.

**Comment by andreabadesso** [comment]
## Continuous Deployment

I agree with using AWS CodePipeline, it makes sense for me to use it for deployments and GithubActions for tasks that are better to stay publicly accessible

## Monitoring

> Whether the wallet-service is being properly synced with the full node in realtime

Our `getLatestBlock` query is now much faster after we refactored it to use the index it has on `height`, I think the /metrics endpoint would be the most recommended option although I agree that it might not be the most scalable one since we might want to monitor other metrics that could also be heavy on the database

Given that, I would also go with the PushGateway solution, I saw the potential drawbacks from using it but since our service is a lambda application, I think they don't apply here

> A reorg is detected with more than 1000 blocks difference

I don't think we need to send the log message alongside the alert. If our mainnet reorgs more than 1000 blocks, we necessarily need to investigate it manually, so the alert will just be the trigger for that manual action, we can read the logs on CloudWatch

> More than X minutes/seconds without a new block from the connected fullnode
> Isn't this solved by the alert in WalletService sync with full-nodes?

Yes it is

> Websocket connection lost with the full-node after X retries

I agree, we can just monitor the logs for alert strings

> Error on Wallet Service connection
> Isn't this solved by alerting on Lambda errors and timeouts?

Yes

**Comment by luislhl** [comment]
> Given that, I would also go with the PushGateway solution, I saw the potential drawbacks from using it but since our service is a lambda application, I think they don't apply here

@andreabadesso , just for clarity, you think we should start with the PushGateway solution?

Also, which potential drawbacks are you mentioning exactly? I didn't understand what are they and why they don't apply.

> I don't think we need to send the log message alongside the alert. If our mainnet reorgs more than 1000 blocks, we necessarily need to investigate it manually, so the alert will just be the trigger for that manual action, we can read the logs on CloudWatch

Great, then alerts on logs seems like a good solution.

**Comment by luislhl** [comment]
## Task List
This is the task list so far, based on @andreabadesso 's and mine opinions

### WalletService
- [ ] Configure the rate-limits on endpoints through Serverless
- [ ] Make sure we are either exiting the Lambda or logging a message containing specific strings when some of those events happen: (we can include other events we think of when doing this)
  - Error on balance calculation
  - Error on MySQL connection
  - Error on FullNode connection

### Daemon
- [ ] Make sure we are logging when the Daemon loses connection to the FullNode, and include an "ALERT" string on the log message
- [ ] Make sure we are logging when a reorg greater than 1000 is detected, and include an "ALERT" string on the log message

### Infra
- [ ] Send all Kubernetes logs to CloudWatch
- [ ] Configure CodePipeline to deploy the WalletService and Daemon
- [ ] Create a Metric Filter and Alarm in CloudWatch to send alerts to Slack when the Daemon logs contains the string "ALERT"
- [ ] Create an Alarm in CloudWatch for when either a Lambda exits or contains specific error string in the logs
- [ ] Create an Alarm in CloudWatch for when the `onNewTxRequest` Lambda is not called in for too much time.

## Under Discussion
- We are still discussing if we will integrate WalletService with Prometheus from the start, or as a future improvement

## Future Tasks
### Infra
- [ ] How to include the log message into Alerts?

**Comment by pedroferreira1** [comment]
> Continuous Deployment

I agree about using the AWS Pipeline, I was just wondering about having our own runners. We currently have some private projects, which have limits on running actions and eventually we need to pay for github to continue running. If we think this will become common (having private projects that run github actions), then using our own runners to prevent having this extra cost with github could be a good solution.

I don't think this is relevant now because I don't see us having many private projects in the near future but I would like to know about the cost (monthly cost and configuration/initial time cost) to have our own runners.

> In a first moment, use just AWS to monitor errors in Lambdas and invocations of onNewTxRequest to make sure the sync is working. This already provides us with a good level of reliability.

The daemon that calls `onNewTxRequest`, right? We might have problem receiving new blocks between the full node and the daemon and between the daemon and the wallet service, right? Maybe we should have a way to compare the data from the 3 services and check if they match by a delta difference (in the height I would say 1 is the limit).

I agree about using PushGateway but in the meantime we could use the prometheus that the full node already saves and create a prometheus for the daemon and create a periodic lambda task that compare the value in both prometheus height with the value in the wallet service db and alert any change > 1. Not sure if I understood correctly the alert need but I think this would be enough.

> A reorg is detected with more than 1000 blocks difference

I think any reorg > 1 should be alerted. At least for now, we can adjust this later.

> So, the idea for now is to just have them configured in API Gateway through Serverless.

Do we have an issue for that?

**Comment by luislhl** [comment]
> I don't think this is relevant now because I don't see us having many private projects in the near future but I would like to know about the cost (monthly cost and configuration/initial time cost) to have our own runners.

Can we create a new issue to make this estimation? I'm not sure this is trivial to do.

We would have to investigate the most cost-effective way of deploying those runners. EC2 is certainly not the best option for cost, since it would be idle for long periods. But I don't know if their runner supports Serverless deployments, or even deploying inside Kubernetes.

> The daemon that calls onNewTxRequest, right? We might have problem receiving new blocks between the full node and the daemon and between the daemon and the wallet service, right? Maybe we should have a way to compare the data from the 3 services and check if they match by a delta difference (in the height I would say 1 is the limit).

My premise to try to simplify things was that checking that new blocks are arriving between the daemon <-> wallet service was enough to check if everything is working. If there is some problem between the full-node <-> daemon, daemon <-> wallet service will also be affected, and we would detect the problem.

> I agree about using PushGateway but in the meantime we could use the prometheus that the full node already saves and create a prometheus for the daemon and create a periodic lambda task that compare the value in both prometheus height with the value in the wallet service db and alert any change > 1. Not sure if I understood correctly the alert need but I think this would be enough.

I think the point above applies here too. The daemon is an intermediary between the full-node and WalletService. If we are able to check if full-node and WalletService are in the same block, that's enough to guarantee that the daemon is also working. So I don't think we need the daemon to export Prometheus metrics.

But the Lambda idea seems good to me. It doesn't need to go on the DB, it would be enough to just get the full-node's metrics (or even call some API in the full-node to get the last block) and call the `getLatestBlock` Lambda in the WalletService, then compare them.

**Comment by pedroferreira1** [comment]
> Can we create a new issue to make this estimation? I'm not sure this is trivial to do.

Yeah, sure. We can start with the AWS pipeline if it's easy to configure. I am not sure how many private projects we intend to have.

> My premise to try to simplify things was that checking that new blocks are arriving between the daemon <-> wallet service was enough to check if everything is working. If there is some problem between the full-node <-> daemon, daemon <-> wallet service will also be affected, and we would detect the problem.

I think it makes sense. I was thinking about monitoring and already returning the point of failure. If we see a problem in the wallet service db we still need to debug if the problem was in the daemon <> full node or in the daemon <> wallet service. But I agree with you, for both situations it makes sense.

**Comment by luislhl** [comment]
@msbrogli , mentioning you here just so you receive a notification.

**Comment by msbrogli** [comment]
Questions:

1) Why does GitHub need access to the database?

2) In regards to "Give Github Actions AWS permissions to update the Security Group", I'm more concerned about the automated task that can give access to others than with the access itself.

3) I like the "Use Self-hosted runners" approach. It can run inside our VPS, which would reduce the permission issues in general.

4) I like the "Use AWS CodePipeline instead of Github Actions" as well. I'm not sure how complex it is to integrate both tho.

5) @luislhl I like your suggestion to use AWS CodePipeline. I'm not sure what would be the impact in our repositories and it seems an important piece of information to make a decision.

6) Monitoring: the ideas seem good to me.

7) Daemon: I agree that we should send the logs to the CloudWatch (or any other centralized log system).

8) Rate limits: But what will be the initial simple rate limits? I think we should have both a limit per IP and a global limit.

9) As far as I understood, the PushGateway would be used to push metrics to Prometheus, right? In this case, how would the metrics be collected? Would it be stored in an S3 file that would be used by PushGateway later? I didn't understand the whole picture.

10) In regards to reorgs, we will have alerts for them in our full nodes. So, maybe the wallet service could benefit from it (instead of creating its own solution). What do you think?

**Comment by luislhl** [comment]
@msbrogli 

> Questions:
> 
> 1. Why does GitHub need access to the database?

Depends on which option we choose to implement. In case we go with AWS CodePipeline, for example, it doesn't.

The initial example was just to illustrate the problem.

> 2. In regards to "Give Github Actions AWS permissions to update the Security Group", I'm more concerned about the automated task that can give access to others than with the access itself.

That's why we would have to be concerned about who has access to the repo. Only people with write access to the repo can trigger the pipeline.

> 3. I like the "Use Self-hosted runners" approach. It can run inside our VPS, which would reduce the permission issues in general.

Exactly, but it seems more complex than using AWS CodePipeline, which can also run inside the VPC.

> 4. I like the "Use AWS CodePipeline instead of Github Actions" as well. I'm not sure how complex it is to integrate both tho.

As far as I researched, CodePipeline can be triggered by actions in Github repos, like on new commits, releases, etc.

> 5. @luislhl I like your suggestion to use AWS CodePipeline. I'm not sure what would be the impact in our repositories and it seems an important piece of information to make a decision.

What kind of impact you are talking about exactly?

I think the only impact is that we would have to give it OAuth access to our Github accounts/repos. Like described in https://docs.aws.amazon.com/codepipeline/latest/userguide/connections-github.html

> 8. Rate limits: But what will be the initial simple rate limits? I think we should have both a limit per IP and a global limit.

Will describe them better. We have to check if API Gateway support limits per IP.

> 9. As far as I understood, the PushGateway would be used to push metrics to Prometheus, right? In this case, how would the metrics be collected? Would it be stored in an S3 file that would be used by PushGateway later? I didn't understand the whole picture.

The metrics would be pushed to PushGateway by Lambdas. PushGateway aggregates them, and makes the result available in a endpoint. Prometheus then pulls them from this endpoint.

It's basically an intermediary to allow pushing metrics instead of the usual pull-based approach in Prometheus.

> 10. In regards to reorgs, we will have alerts for them in our full nodes. So, maybe the wallet service could benefit from it (instead of creating its own solution). What do you think?

I need to understand better how those alerts will work in the full nodes. Also, @andreabadesso 's opinion is important here.

**Comment by msbrogli** [comment]
@luislhl 

1) About why GitHub needs access, I meant the reason. But I guess I got it. Just to check: GitHub needs access to be able to run the migrations. Is that right?

2) Another idea is to use GitHub to trigger the execution of the migrations in the Kubernetes. But it might be even easier using the AWS CodePipeline.

3) About the impact in the repositories, I meant to move all actions to AWS CodePipeline. For instance, how would a project configure the AWS CodePipeline? Does it support to read the configuration file directly from the repository? Would we need to use TerraForm?

4) If API Gateway does not support throttling per IP, we can use it for the global limits. But I think we should have another intermediary to throttle per IP (e.g., nginx).

5) Can't Prometheus pull from S3?

6) We still need to develop the reorg alerts in the full node. But it is something that we will do anyway. It is very important to learn as soon as possible when a reorg occurs. But I agree we need to put @andreabadesso in this loop.

**Comment by luislhl** [comment]
@msbrogli 

> 1. About why GitHub needs access, I meant the reason. But I guess I got it. Just to check: GitHub needs access to be able to run the migrations. Is that right?

Yes, to run the migrations. But if we choose another solution, it's not needed.

> 2. Another idea is to use GitHub to trigger the execution of the migrations in the Kubernetes. But it might be even easier using the AWS CodePipeline.

This raises other concerns, like giving access on the Kubernetes API to GitHub actions. This API is currently blocked by IP actually, for security reasons. We would need to expose it. Or we would need a indirect form of triggering jobs in Kubernetes, like sending a task to a queue or something similar.

So it seems CodePipeline would be easier.

> 3. About the impact in the repositories, I meant to move all actions to AWS CodePipeline.

The idea is not to replace GitHub Actions entirely, just for deployment. I would let PR validation tasks and some other tasks (like building Docker images for hathor-core) still in Github Actions.

> For instance, how would a project configure the AWS CodePipeline? Does it support to read the configuration file directly from the repository? Would we need to use TerraForm?

Those are good questions, I still don't have answer for them. But will try to figure them out.

UPDATE:
It supports reading the configuration file directly from the repository.

The way we configure it is similar to Github Actions, with a file that decribes the steps, but with a little more bureaucracy to setup connection with GitHub and integration with AWS CodeBuild, which is a tool used in conjunction with AWS CodePipeline to run the builds.

All of this bureaucracy can be declared and created using Terraform. I think we should use it, since we used it for everything we are creating for WalletService.

> 5. Can't Prometheus pull from S3?

I don't think so, and a quick search didn't return anything showing this is possible. We would need a service that reads from S3 and exports metrics in an API. Maybe there is some project in the internet with such service that we would just have to deploy.

**Comment by luislhl** [comment]
I have created a PR adding a .md file with the condensed information already discussed here.
https://github.com/HathorNetwork/hathor-wallet-service/pull/111

I think we should continue discussions there, so I'll be closing this issue.
